execution/function/scalar.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package function
+
+import (
+	"context"
+	"math"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/thanos-community/promql-engine/execution/model"
+)
+
+type scalarFunctionOperator struct {
+	stepsBatch  int
+	currentStep int64
+	mint        int64
+	maxt        int64
+	step        int64
+	pool        *model.VectorPool
+	next        model.VectorOperator
+}
+
+func (o *scalarFunctionOperator) Explain() (me string, next []model.VectorOperator) {
+	return "[*scalarFunctionOperator]", []model.VectorOperator{}
+}
+
+func (o *scalarFunctionOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	return nil, nil
+}
+
+func (o *scalarFunctionOperator) GetPool() *model.VectorPool {
+	return o.pool
+}
+
+func (o *scalarFunctionOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+	in, err := o.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(in) == 0 {
+		return nil, nil
+	}
+
+	result := o.GetPool().GetVectorBatch()
+	for _, vector := range in {
+		sv := o.GetPool().GetStepVector(vector.T)
```

Comment:
I noticed that in the original implementation we modify the input vector in place: https://github.com/thanos-io/promql-engine/pull/265/files#diff-45b283ce3f8010b8cd1c8c50b568e54e58a974d8b4ba75b14e8821680641b0fcL219.

Can we do the same here?

Reply:
i think we could do that yes

Reply:
note: tried and it did fail some range query tests; decided to go back to using the pool from the scalar operator itself.